nemo/collections/asr/models/ctc_bpe_models.py

```diff
@@ -0,0 +1,202 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+from dataclasses import dataclass
+from typing import Dict, Optional
+
+import torch
+from omegaconf import MISSING, DictConfig, ListConfig, OmegaConf
+
+from nemo import logging
+from nemo.collections.asr.data.audio_to_text import AudioToBPEDataset
+from nemo.collections.asr.metrics.wer_bpe import WERBPE
+from nemo.collections.asr.models.ctc_models import EncDecCTCModel, EncDecCTCModelConfig
+from nemo.collections.asr.parts.features import WaveformFeaturizer
+from nemo.collections.asr.parts.perturb import process_augmentations
+from nemo.collections.common import tokenizers
+from nemo.core.neural_types import *
+from nemo.utils.decorators import experimental
+
+__all__ = ['EncDecCTCModelBPE', 'JasperNetBPE', 'QuartzNetBPE']
+
+
+@dataclass
+class EncDecCTCModelBPEConfig(EncDecCTCModelConfig):
+    tokenizer: DictConfig = MISSING
+
+
+@experimental
+class EncDecCTCModelBPE(EncDecCTCModel):
+    """Encoder decoder CTC-based models with Byte Pair Encoding."""
+
+    def __init__(self, cfg: EncDecCTCModelBPEConfig, trainer=None):
+        self.tokenizer_cfg = OmegaConf.to_container(cfg.tokenizer, resolve=True)  # type: dict
+        self.tokenizer_path = self.tokenizer_cfg.pop('path')  # Remove path and resolve based on tokenizer type
+
+        if os.path.exists(os.path.join(self.tokenizer_path, 'tokenizer.model')):
+            # This is a BPE Tokenizer
+            model_path = os.path.join(self.tokenizer_path, 'tokenizer.model')
```

Comment:
I think you should have the user specify the tokenizer class they want to use as opposed to choosing based on the name of the file.

Reply:
Using classpath seems to be frowned upon by rest of the team, and I editing classpath inside command line is tedious. Its preferred to simply provide a unified interface - pass a directory with a vocab / model file and we can use it. Its particularly useful to speed test multiply different tokenizers with different number of tokens by just switching path to directory.

What we should do is log the name of the tokenizer and its parameters so it is clear 1) the tokenizer was successfully loaded 2) what type the tokenizer is.

Reply:
why is 'tokenizer.model' hard-coded? This is tokenizer's model, right? Users will want to name it differently, e.g. bpe1024.model, wp4096.model, etc.

Reply:
This is the output of the prepare_asr_tokenizer.py script below. Its generated by sentencepiece library, so if the user wants different tokenizers, they will have different output directories from the below script. We are pointing more to a directory than to the script itself. Since sentence piece is the special one out of the 2 supported ones - WPE and sentencepiece, and only sentence piece generates tokenizer.model, I do an explicit check.

Reply:
I didn't mean a `cls`, it could just be a string that accepts either `SentencePiece` or `NeMoBert` or something along those lines

Reply:
Ok that sounds like a good idea. Lets use `type: <name>` where <name> = either `BPE` ~ SentencePiece and `WPE` ~ NeMoBERTTokenizer resolution. Error if any other name is provided.

Reply:
@blisc Done. Now we have `tokenizer.dir` to point to the directory and `tokenizer.type` with the above two values to ease the logic switching between tokenizers.